### DISCUSSION TITLE
Emporia API changed - Upgrade to 1.3.0 required

### DISCUSSION THREAD

**Discussion by jertel** [jertel/vuegraf#58] [General]
Outage in effect as of 12pm EDT on Aug 18, 2021 and will remain in affect until you upgrade Vuegraf.

Base library has been adjusted as mentioned here: https://github.com/magico13/PyEmVue/issues/25

Upgrade to Vuegraf 1.3.0 to resolve this outage.

**Comment by prophead** [comment]
I have a stable running vuegraf under python under systemd, no docker (well, until this morning).  Can you give some help on how to upgrade from the command line?  TIA

**Reply by jertel** [reply]
Docker is the easier solution, so when you said "well until this morning" do you intend to start using Docker?

**Reply by prophead** [reply]
no docker, systemd just starts python3 running vuegraf, rock solid.  I hate docker.  Docker is not easier.  I don't want to argue that point, I just want to know the command to upgrade.

**Reply by jertel** [reply]
`git pull`? Not sure how you installed it to begin with so I can't give much help on how to upgrade it.

**Reply by prophead** [reply]
ok, that updates the core files, how do I update the pyemvue library?

**Reply by jertel** [reply]
Same command as shown on the [readme](https://github.com/jertel/vuegraf):
 ```
pip install -r src/requirements.txt
```

**Comment by cooldil** [comment]
Nested devices aren't being picked up...

Also need to disregard the ChannelNumber/ChannelName with a value of "TotalUsage" along with "Balance" that you are already doing.  TotalUsage shows up for Vues that have Solar enabled.

**Reply by jertel** [reply]
Can you give me an example of your nested setup? I don't have anything nested so I'll see about supporting it but will need help testing.

**Reply by jertel** [reply]
I've attempted to support nested devices, as well as ignore TotalUsage devices. This is now available in the latest Docker image and latest Python code. What I'm not sure about is whether there will be issues with usage summation due to the parent device also being recording into influxDB. Let me know if this solves it and the data looks correct. If so I'll create a point release.

**Reply by cooldil** [reply]
My setup has a total of three Vue2 units and an Outlet.  One of the Vues is nested underneath another Vue.  The outlet is then nest underneath one of the "channels" of the nested Vue.

```
Unit 1
   |__Channel 1
   |__Channel ...
   |__Channel 16
Unit 2
   |__Channel 1
   |__Channel ...
   |__Channel 16
   |__Unit 3
           |__ Channel 1
           |__ Channel 2
           |__ Channel 3
                      |__ Outlet
           |__ Channel 4
           |__ Channel ...
           |__ Channel 16

```         

Run the code below using the PyEmVue library directly to validate the output of the get_device_list_usage function:


```
import datetime
import json
import time
from pyemvue import PyEmVue
from pyemvue.enums import Scale, Unit

#Function to print out the devices and channels
def print_recursive(usage_dict, info, scaleBy=1, unit='kWh', depth=0):
    for gid, device in usage_dict.items():
        for channelnum, channel in device.channels.items():
            name = channel.name
            if name == 'Main':
                name = info[gid].device_name
            usage = channel.usage or 0
            print('-'*depth, f'{gid} {channelnum} {name} {usage*scaleBy} {unit}')
            if channel.nested_devices:
                print_recursive(channel.nested_devices, info, scaleBy=scaleBy, unit=unit, depth=depth+1)

#Login using tokens 
vue = PyEmVue()
#vue.login(username='you@email.com', password='password', token_storage_file='keys.json')
with open('keys.json') as f:
    data = json.load(f)
vue.login(id_token=data['idToken'],
    access_token=data['accessToken'],
    refresh_token=data['refreshToken'],
    token_storage_file='keys.json')


now = datetime.datetime.utcnow()
devices = vue.get_devices()
deviceGids = []
deviceInfo = {}
for device in devices:
    if not device.device_gid in deviceGids:
        deviceGids.append(device.device_gid)
        deviceInfo[device.device_gid] = device
        print(device.device_gid, device.manufacturer_id, device.model, device.firmware)
        for chan in device.channels:
            print('\t', chan.device_gid, chan.name, chan.channel_num, chan.channel_multiplier)
    else:
        deviceInfo[device.device_gid].channels += device.channels

device_usage_dict = vue.get_device_list_usage(deviceGids, now, scale=Scale.HOUR.value, unit=Unit.KWH.value)
print_recursive(device_usage_dict, deviceInfo)

```

Output without nesting (I have zeroed out my usage values)

```
 00003 1,2,3 Unit 3 0.00 kWh
 00003 1 U3C1 0.00 kWh
 00003 2 U3C2 0.00 kWh
 00003 3 U3C3 0.00 kWh
 00003 4 U3C4 0.00 kWh
 00003 5 U3C5 0.00 kWh
 00003 6 U3C6 0.00 kWh
 00003 7 U3C7 0.00 kWh
 00003 8 U3C8 0.00 kWh
 00003 9 U3C9 0 kWh
 00003 10 U3C10 0 kWh
 00003 11 U3C11 0.00 kWh
 00003 12 U3C12 0.00 kWh
 00003 13 U3C13 0 kWh
 00003 14 U3C14 0.00 kWh
 00003 15 U3C15 0.00 kWh
 00003 16 U3C16 0 kWh
 00003 Balance Balance 0.00 kWh
 00004 1,2,3 Outlet 0.00 kWh
 00001 1,2,3 Unit 1 0.00 kWh
 00001 1 U1C1 0.00 kWh
 00001 2 U1C2 0.00 kWh
 00001 3 U1C3 0.00 kWh
 00001 4 U1C4 0.00 kWh
 00001 5 U1C5 0.00 kWh
 00001 6 U1C6 0.00 kWh
 00001 7 U1C7 0.00 kWh
 00001 8 U1C8 0.00 kWh
 00001 9 U1C9 0 kWh
 00001 10 U1C10 0 kWh
 00001 11 U1C11 0 kWh
 00001 12 U1C12 0 kWh
 00001 13 U1C13 0 kWh
 00001 14 U1C14 0 kWh
 00001 15 U1C15 0 kWh
 00001 Balance Balance 0.00 kWh
 00002 1,2,3 Unit 2 -0.00 kWh
 00002 1 U2C1 0 kWh
 00002 2 U2C2 0 kWh
 00002 3 U2C3 0 kWh
 00002 4 U2C4 0.00 kWh
 00002 5 U2C5 0.00 kWh
 00002 6 U2C6 0.00 kWh
 00002 7 U2C7 0.00 kWh
 00002 8 U2C8 0.00 kWh
 00002 9 U2C9 0 kWh
 00002 10 U2C10 0 kWh
 00002 11 U2C11 0 kWh
 00002 12 U2C12 0 kWh
 00002 13 U2C13 0 kWh
 00002 14 U2C14 0.00 kWh
 00002 15 U2C15 0.00 06 kWh
 00002 16 U2C16 0 kWh
 00002 TotalUsage TotalUsage 0.00 kWh
 00002 Balance Balance -0.00 kWh
```

Output with a nested Vue (0003) and nested Outlet (0004)

```
 00001 1,2,3 Unit 1 0.00 kWh
 00001 1 U1C1 0.00 kWh
 00001 2 U1C2 0.00 kWh
 00001 3 U1C3 0.00 kWh
 00001 4 U1C4 0.00 kWh
 00001 5 U1C5 0.00 kWh
 00001 6 U1C6 0.00 kWh
 00001 7 U1C7 0.00 kWh
 00001 8 U1C8 0.00 kWh
 00001 9 U1C9 0 kWh
 00001 10 U1C10 0 kWh
 00001 11 U1C11 0 kWh
 00001 12 U1C12 0 kWh
 00001 13 U1C13 0 kWh
 00001 14 U1C14 0 kWh
 00001 15 U1C15 0 kWh
 00001 Balance Balance 0.00 kWh
 00002 1,2,3 Unit 2 -0.00 kWh
 - 00003 1,2,3 Unit 3 0.00 kWh
 - 00003 1 U3C1 0.00 kWh
 - 00003 2 U3C2 0.00 kWh
 - 00003 3 U3C3 0.00 kWh
 -- 00004 1,2,3 Outlet 0.00 kWh
 - 00003 4 U3C4 0.00 kWh
 - 00003 5 U3C5 0.00 kWh
 - 00003 6 U3C6 0.00 kWh
 - 00003 7 U3C7 0.00 kWh
 - 00003 8 U3C8 0.00 kWh
 - 00003 9 U3C9 0 kWh
 - 00003 10 U3C10 0 kWh
 - 00003 11 U3C11 0.00 kWh
 - 00003 12 U3C12 0.00 kWh
 - 00003 13 U3C13 0 kWh
 - 00003 14 U3C14 0.00 kWh
 - 00003 15 U3C15 0.00 kWh
 - 00003 16 U3C16 0 kWh
 00002 1 U2C1 0 kWh
 00002 2 U2C2 0 kWh
 00002 3 U2C3 0 kWh
 00002 4 U2C4 0.00 kWh
 00002 5 U2C5 0.00 kWh
 00002 6 U2C6 0.00 kWh
 00002 7 U2C7 0.00 kWh
 00002 8 U2C8 0.00 kWh
 00002 9 U2C9 0 kWh
 00002 10 U2C10 0 kWh
 00002 11 U2C11 0 kWh
 00002 12 U2C12 0 kWh
 00002 13 U2C13 0 kWh
 00002 14 U2C14 0.00 kWh
 00002 15 U2C15 0.00 06 kWh
 00002 16 U2C16 0 kWh
 00002 TotalUsage TotalUsage 0.00 kWh
 00002 Balance Balance -0.00 kWh
```

Think within the for loop starting on line 193, need to check for the "nested_devices" property and account for the devices/channels...

I can certainly help with any testing...

**Reply by cooldil** [reply]
> I've attempted to support nested devices, as well as ignore TotalUsage devices. This is now available in the latest Docker image and latest Python code. What I'm not sure about is whether there will be issues with usage summation due to the parent device also being recording into influxDB. Let me know if this solves it and the data looks correct. If so I'll create a point release.

Unfortunately, based on what I am seeing, nesting *will* impact the summation.  The parent values will be reflected AFTER subtracting the value/s from the nested items.  For example if you have a nested Outlet consuming 1kW on a circuit that has a TOTAL load of 3kW, the load reported by PyEmVue (or the app for that matter), will be 1kW for the Outlook and 2kW (3 - 1) for the Circuit.

**Reply by jertel** [reply]
The way I'm reading this it sounds like this is what I would want to happen. E.g., I configured the nesting deliberately so I am expecting Emporia to perform the subtraction and reflect those adjusted values in their API. I'm curious what you think of the current version I pushed this morning. So when you get a chance please give it a shot. If you add some real values generated by Vuegraf next to that nesting output you provided above it will help me see what you're seeing.

Thanks!

**Reply by cooldil** [reply]
New version is erroring out when it hits the nested device due to the device not having a "channels" attribute.  Error occurs regardless if the nested device is a Vue or the Outlet...  Haven't yet debugged in much detail yet to determine why the recursion function causes the error when the initial entry doesn't.... 


```
2021-08-19 16:29:36.985262 | ERROR | Failed to record new usage data: (<class 'AttributeError'>, AttributeError("'int' object has no attribute 'channels'"), <traceback object at 0x06EA71E8>)
Traceback (most recent call last):
  File "jertel.py", line 218, in <module>
    extractDataPoints(device, usageDataPoints)
  File "jertel.py", line 118, in extractDataPoints
    extractDataPoints(nestedDevice, usageDataPoints)
  File "jertel.py", line 114, in extractDataPoints
    for chanNum, chan in device.channels.items():
AttributeError: 'int' object has no attribute 'channels'
```

**Reply by jertel** [reply]
Thanks. I've pushed a fix.

**Reply by cooldil** [reply]
Seems to have done the trick!  Thanks!

**Reply by cooldil** [reply]
> The way I'm reading this it sounds like this is what I would want to happen. E.g., I configured the nesting deliberately so I am expecting Emporia to perform the subtraction and reflect those adjusted values in their API. 
> 
> Thanks!

Clarification on how the usage is reported for nested devices.  It depends on the API call being made to pull the usage.  For the get_chart_usage API call, the parent will report the TOTAL load (and not the difference between the total and nested device/s).  For the get_device_list_usage API call, the parent will report the load after subtracting the nested device/s load (total - nested device/s).

In other words, the detailed measurements in your project would have the TOTAL load while the non-detailed will have the calculated load. :-)

**Reply by kpnobvious** [reply]
I too am having this issue and will try the fix thanks Jertel!

**Reply by kpnobvious** [reply]
> 
> 
> I too am having this issue and will try the fix thanks Jertel!

i updated to the latest version and that picked up my 2 nested smart plugs

**Reply by jertel** [reply]
Thanks for the confirmation. I'll create a point release even though there's a reported issue with detailed per-second data not being calculated correctly for nested devices. This inconsistency really complicates things!

**Reply by kpnobvious** [reply]
Yes it does.  I don't pull detailed data at this time.  I would like to change the script to run less frequently and pull data over longer time periods.
I think by default it pulls 1 minute of data points every 60 seconds (intervalSecs).  If i wanted to pull say 15 minutes of per minute data points every 15 minutes i would just have to update intervalSecs.  i have to try that.

**Reply by jertel** [reply]
Yes, the config key is `updateIntervalSecs` so set that to 900 for a 15 minute interval. And if you want to disable detailed data pulls set `detailedIntervalSecs` to some impossibly high interval, such as 3153600000 (100 years).

Would look like this in your config:

```
   ...
    "updateIntervalSecs": 900,
    "detailedIntervalSecs": 3153600000,
    "accounts": [ 
    ...
```

**Comment by Cjkeenan** [comment]
I just noticed this new update and as a result, I have a large gap in my database now. Is there any way to backfill that data? 

Also has there been any update to the odd errors regarding the docker image?

Finally, regarding the new update, it does not really matter in my case since I only have a single device, but my "device_name" attribute changed, which could be very annoying if I had plug or per circuit data. Not sure if this is a bug or if there is any way to remediate this.
